### PR TITLE
store: allow boosted downloads

### PR DESCRIPTION
**Pull request by jsign**
Includes an experimental flag `--boost-download` which opens multiple TCP connections in content-range compatible HTTP endpoints. If the endpoint isn't compatible, it will fall back to doing usual downloads.

Leaving this as a PR until I've some feedback from a storage provider doing some tests and confirming.

### CODE REVIEW

File: service/store/store.go [hunk 1]
```diff
@@ -505,11 +510,18 @@ func (s *Store) WriteDealData(b *Bid) (string, error) {
...
-	if err != nil {
-		return "", fmt.Errorf("parsing data uri: %w", err)
+	carDownloadPath := s.dealDataFilePathFor(bidID, payloadCid)
+
+	if s.boostedDownload {
+		ok := s.boostDownload(s.ctx, uri, carDownloadPath)
+		if ok {
+			return carDownloadPath, nil
+		}
 	}
```
Comment (jsign) [inline service/store/store.go:520]: As an experimental feature, trying to keep it separate.
If the flag has been enabled, we try using this download style. If for any reason things didn't worked out, we fallback with the usual style.

What's outside this if case is the normal download style.

**Comment by jsign** [comment]
Two storage providers tested this experimental feature and looking quite good on their side, so more comfortable with formally merging it after a review.

File: service/store/store.go [hunk 2]
```diff
@@ -523,14 +535,32 @@ func (s *Store) WriteDataURI(bidID auction.BidID, payloadCid, uri string, size u
...
+	if err != nil {
+		return "", fmt.Errorf("parsing data uri: %w", err)
+	}
 	if err := duri.Write(ctx, f); err != nil {
 		return "", fmt.Errorf("writing data uri %s: %w", uri, err)
 	}
 	return f.Name(), nil
 }
 
+func (s *Store) boostDownload(ctx context.Context, uri string, carDownloadPath string) bool {
```
Comment (brunocalza) [inline service/store/store.go:549]: Going to throw some stylish/naming-conventions comments just for you to get a sense of how I usually think about code 

- I have a preference towards the `error` returning type to indicate a failed operation and not the `bool`. It just seems more natural and go's idiomatic.  

- For methods that return `bool` (predicate functions), I prefer to use the naming convention that indicates that the method returns a boolean: yes/no question-like method names (e.g. `IsReady`, `HasPermissions`). (This comment is unrelated to the first)

**Review by brunocalza — APPROVED** [review]
PR seems fine and ready to be merged. Some observations

- I think the [points you made in Slack](https://textile-public.slack.com/archives/C02D4N98QQ4/p1636132960301700) are really good to have in mind. Unfortunately, I don't have too much to add 
- Did some observations (not a blocker) in the definition of the `boostDownload` method.

File: service/store/store.go (hunk 2 above)
Comment (jsign) [reply service/store/store.go:549]: I agree with both conventions for general things. The `ok` return is mostly a simplification since probably the error probably won't evolve in the future, so simplifying creating a specific error type here. But highly agree that _controlled errors_ are probably better be typed errors that should be checked by the caller (and explained in method's comment).

Also agree on naming for methods that return bools, as you mentioned, this isn't exactly the case since the method isn't answering a "question", but is mostly taking a shortcut to signal a "failed" operation.

In summary, agree on the convention, and feels to me we should potentially try to stick with that moving forward. I think is a convention that isn't enforced today, but we could start migration whenever we come up with places that need that migration.

**Comment by jsign** [comment]
> PR seems fine and ready to be merged. Some observations
> 
> * I think the [points you made in Slack](https://textile-public.slack.com/archives/C02D4N98QQ4/p1636132960301700) are really good to have in mind. Unfortunately, I don't have too much to add
> * Did some observations (not a blocker) in the definition of the `boostDownload` method.

Thanks @brunocalza  for the review!
